routes/error-tracker.js

```diff
@@ -0,0 +1,219 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview
+ * Handle error requests from clients and log them.
+ */
+
+const express = require('express');
+const logging = require('@google-cloud/logging');
+const winston = require('winston');
+const statusCodes = require('http-status-codes');
+const url = require('url');
+const router = new express.Router();
+const appEngineProjectId = 'amp-error-reporting';
+const logName = 'javascript.errors';
+const SERVER_START_TIME = Date.now();
+const filteredMessageOrException = ['stop_youtube',
+  'null%20is%20not%20an%20object%20(evaluating%20%27elt.parentNode%27)'];
+
+/**
+ * @enum {int}
+ */
+const SEVERITY = {
+  INFO: 200,
+  ERROR: 500,
+};
+
+/**
+ * @param {string} message
+ * @param {string} exception
+ * @return {boolean}
+ */
+function isFilteredMessageOrException(message, exception) {
+  return filteredMessageOrException.some(function(msg) {
+    return message.includes(msg) || exception.includes(msg);
+  });
+}
+
+/**
+ * @desc extract params in GET request from query and fill errorEvent object
+ * log level by default is INFO.
+ * @param {httpRequest} req
+ * @param {response} res
+ * @param {middleware} next
+ */
+function getHandler(req, res, next) {
+  const params = req.query;
+  if (params.m === '' && params.s === '') {
+    res.status(statusCodes.BAD_REQUEST);
+    res.send({error: 'One of \'message\' or \'exception\' must be present.'});
+    res.end();
+    winston.log('Error', 'Malformed request: ' + params.v.toString(), req);
+    return;
+  }
+  const referer = params.r;
+  let errorType = 'default';
+  let isUserError = false;
+  if (params.a === '1') {
+    errorType = 'assert';
+    isUserError = true;
+  }
+
+  // if request comes from the cache and thus only from valid
+  // AMP docs we log as "Error"
+  let severity = SEVERITY.INFO;
+  let isCdn = false;
+  if (referer.startsWith('https://cdn.ampproject.org/') ||
+    referer.includes('.cdn.ampproject.org/') ||
+    referer.includes('.ampproject.net/')) {
+    severity = SEVERITY.ERROR;
+    errorType += '-cdn';
+    isCdn = true;
+  } else {
+    errorType += '-origin';
+  }
```

Comment:
Yeah. All test cases that do not have referrer set to any of the cdn or amp project links.